### PR TITLE
Acceptance tests for create new investment project

### PR DESCRIPTION
**Pull request by venureddy95**
1. Added a page objects file for Creating new Investment project
2. Added a feature file for Creating new Investment project with 8 scenarios
3. Added a step definitions file for Creating new Investment project

### CODE REVIEW

File: test/acceptance/page_objects/InvestmentProject.js [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+          .enterProjectName(projectName)
+          .enterDescription()
+          .clickProjectBeSharedYes()
+          .clicksignedaNDANo()
+          .enterLandMonth()
+          .enterLandYear()
+        .submitForm('form')
+      },
+
+      createNewInvestmentProjectWithDifferentClientRelationManager (projectName) {
```
Comment (teneightfive) [inline test/acceptance/page_objects/InvestmentProject.js]: There seems a lot of repetition between this method and the previous one.

Could we have an option to pass a parameter in to this function which sets whether the client relationship manager is yes or no rather?

Comment (venureddy95) [inline test/acceptance/page_objects/InvestmentProject.js]: @teneightfive : Once we select Client relationship Manager to NO, there is an additional field that populates. Which is handled in the below Step definitions file. So that the reason I kept if separately.
When(/^I create a new Investment project with a different client relation manager$/, async () => {

File: test/acceptance/features/step_definitions/create-a-investment-project-step-definitions.js [hunk 2]
```diff
@@ -0,0 +1,224 @@
+const faker = require('faker')
+const { client } = require('nightwatch-cucumber')
+const { defineSupportCode } = require('cucumber')
+
+defineSupportCode(({ Given, Then, When }) => {
+  const Company = client.page.Company()
+  const Contact = client.page.Contact()
+  const Investment = client.page.InvestmentProject()
+  const randomCompanyName = `TestingNewCompany-1206-1`
```
Comment (feedmypixel) [inline test/acceptance/features/step_definitions/create-a-investment-project-step-definitions.js]: Its worth reading up on [Template Literals](https://leanpub.com/understandinges6/read/#leanpub-auto-template-literals)

**TL;DR**

use single quotes if just a string literal:
```
const randomCompanyName = 'TestingNewCompany-1206-1'
```

use template literals if you need to substitute a variable:
```
const identifier = '1206-1'
const randomCompanyName = `TestingNewCompany-${identifier}`
```

File: test/acceptance/features/step_definitions/create-a-investment-project-step-definitions.js [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+  const fixedCompanyName = `Sony`
```
Comment (feedmypixel) [inline test/acceptance/features/step_definitions/create-a-investment-project-step-definitions.js]: same can be a string literal

File: test/acceptance/features/step_definitions/create-a-investment-project-step-definitions.js [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+  let projectName = ''
```
Comment (feedmypixel) [inline test/acceptance/features/step_definitions/create-a-investment-project-step-definitions.js]: `projectName ` can be set here globally once? Does it need to be reassigned within each `When`

File: test/acceptance/features/step_definitions/create-a-investment-project-step-definitions.js [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+    await Contact
+      .clickOnFirstCompanyFromList()
+    await Investment
+      .clickInvestmentsTab()
+      .createNewInvestmentProject(projectName)
+  })
+
+  Then(/^I see the Investment project creation confirmation message$/, async () => {
+    await Investment
+      .getText('@projectNameFromSummaryPage', function (result) {
```
Comment (feedmypixel) [inline test/acceptance/features/step_definitions/create-a-investment-project-step-definitions.js]: could use arrow functions here:
```
(result) => {
```

rather than:
```
function (result) {
```

File: test/acceptance/features/step_definitions/create-a-investment-project-step-definitions.js [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+    await Contact
+      .clickOnFirstCompanyFromList()
+    await Investment
+      .clickInvestmentsTab()
+      .getText('@projectNameFromCompanyProfile', function (result) {
+        Investment.assert.equal(result.value, projectName)
+      })
+  })
+
+  Then(/^I verify Type of Investment is shown as "([^"]*)"$/, async (type) => {
```
Comment (feedmypixel) [inline test/acceptance/features/step_definitions/create-a-investment-project-step-definitions.js]: `type` is very generic how about `typeOfInvestment`?

File: test/acceptance/page_objects/InvestmentProject.js [hunk 7]
```diff
@@ -0,0 +1,378 @@
+const faker = require('faker')
+
+module.exports = {
+  url: process.env.QA_HOST,
+  props: {
```
Comment (feedmypixel) [inline test/acceptance/page_objects/InvestmentProject.js]: does `props` need to be here if its empty?

File: test/acceptance/page_objects/InvestmentProject.js [hunk 8]
```diff
@@ -0,0 +1,378 @@
...
+  },
+  elements: {
+    investmentsTab: {
+      selector: "//nav/a[contains(@href, 'investments')]",
+      locateStrategy: 'xpath',
+    },
+    addInvestmentProjectButton: '#main-content div article div a',
+    equitySourceYes: '#client-equity-source-options label:nth-child(2)',
```
Comment (feedmypixel) [inline test/acceptance/page_objects/InvestmentProject.js]: feels like using `label:nth-child(2)` could be quite brittle. did we agree a way of adding `qa-` className or id hooks into the markup? @tyom @teneightfive ?

File: test/acceptance/page_objects/InvestmentProject.js [hunk 9]
```diff
@@ -0,0 +1,378 @@
...
+      selector: "//label[@for='input--is-relationship-manager--1']",
+      locateStrategy: 'xpath',
+    },
+    clientRelationshipManagerNo: {
+      selector: "//label[@for='input--is-relationship-manager--2']",
+      locateStrategy: 'xpath',
+    },
+    clientRelationshipManager: '#client_relationship_manager-wrapper input',
+    clientRelationshipManagerList: '#client_relationship_manager-wrapper ul li:nth-child(4)',
+    // clientRelationsshipManagerName: '#referral_source_adviser-wrapper select option[selected=selected]',
```
Comment (feedmypixel) [inline test/acceptance/page_objects/InvestmentProject.js]: no need to commit commented out code

File: test/acceptance/page_objects/InvestmentProject.js [hunk 10]
```diff
@@ -0,0 +1,378 @@
...
+    sourceCompanySearch: '#search-q',
+    firstCompanyFromList: {
+      selector: ".//*[@id='main-content']/div/article/ol/li[1]//h3",
+      locateStrategy: 'xpath',
+    },
+  },
+  commands: [
+    {
+      clickInvestmentsTab () {
+        return this
```
Comment (feedmypixel) [inline test/acceptance/page_objects/InvestmentProject.js]: is there a need to abstract this? it is only one command

File: test/acceptance/page_objects/InvestmentProject.js [hunk 11]
```diff
@@ -0,0 +1,378 @@
...
+      selectBusinessActivity () {
+        return this
+          .click('@businessActivity')
+          .click('@businessActivityList')
+      },
+      enterProjectName (projectName) {
+        return this
+          .setValue('@projectName', projectName)
+      },
+      enterDescription () {
```
Comment (feedmypixel) [inline test/acceptance/page_objects/InvestmentProject.js]: There is a lot of abstraction in this file that IMO feels like its not needed, I'm not convinced that you gain much from changing:

```
this.setValue('@projectName', projectName)
```

to:
```
this.enterProjectName(projectName)
```

If there are multiple methods then granted it feels worth abstracting

File: test/acceptance/page_objects/InvestmentProject.js (hunk 8 above)
Comment (tyom) [inline test/acceptance/page_objects/InvestmentProject.js]: The new form components provide a number of id hooks that can be used in QA. I'd avoid adding anything new for now. Just need to keep an eye on tests and update them as they break.

File: test/acceptance/features/step_definitions/create-a-investment-project-step-definitions.js [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+    projectName = faker.commerce.productName()
+    await Company
+      .navigate()
+      .findCompany(randomCompanyName)
+    await Contact
+      .clickOnFirstCompanyFromList()
+    await Investment
+      .clickInvestmentsTab()
+      .createNewInvestmentProjectWithDifferentClientRelationManager(projectName)
+    .setValue('@clientRelationshipManager', 'a')
```
Comment (tyom) [inline test/acceptance/features/step_definitions/create-a-investment-project-step-definitions.js]: Inconsistent indentation.

File: test/acceptance/features/step_definitions/create-a-investment-project-step-definitions.js [hunk 13]
```diff
@@ -0,0 +1,224 @@
...
+    projectName = faker.commerce.productName()
+    await Company
+      .navigate()
+      .findCompany(randomCompanyName)
+    await Contact
+      .clickOnFirstCompanyFromList()
+    await Investment
+      .clickInvestmentsTab()
+      .createNewInvestmentProjectWithDifferentReferralSourceAdviser(projectName)
+    .setValue('@referralSourceAdviser', 'a')
```
Comment (tyom) [inline test/acceptance/features/step_definitions/create-a-investment-project-step-definitions.js]: Inconsistent indentation.

File: test/acceptance/features/step_definitions/create-a-investment-project-step-definitions.js [hunk 14]
```diff
@@ -0,0 +1,224 @@
...
+      .findCompany(randomCompanyName)
+    await Contact
+      .clickOnFirstCompanyFromList()
+    await Investment
+      .clickInvestmentsTab()
+      .clickAddInvestmentProjectButton()
+      .clickEquitySourceNo()
+      .submitForm('form')
+      .enterSourceCompanySearch(fixedCompanyName)
+    .submitForm('form')
```
Comment (tyom) [inline test/acceptance/features/step_definitions/create-a-investment-project-step-definitions.js]: Inconsistent indentation.

File: test/acceptance/page_objects/InvestmentProject.js (hunk 10 above)
Comment (tyom) [inline test/acceptance/page_objects/InvestmentProject.js]: I agree with Ben. I fail to see the value of these simple abstractions.

File: test/acceptance/page_objects/InvestmentProject.js [hunk 15]
```diff
@@ -0,0 +1,378 @@
...
+          .clickTypeOfInvestmentCommitmentToInvest()
+          .selectPrimarySector()
+          .selectBusinessActivity()
+          .enterProjectName(projectName)
+          .enterDescription()
+          .clickProjectBeSharedYes()
+          .clicksignedaNDANo()
+          .enterLandMonth()
+          .enterLandYear()
+        .submitForm('form')
```
Comment (tyom) [inline test/acceptance/page_objects/InvestmentProject.js]: Inconsistent indentation.

File: test/acceptance/page_objects/InvestmentProject.js [hunk 16]
```diff
@@ -0,0 +1,378 @@
...
```
Comment (tyom) [inline test/acceptance/page_objects/InvestmentProject.js]: Inconsistent indentation.

File: test/acceptance/page_objects/InvestmentProject.js [hunk 17]
```diff
@@ -0,0 +1,378 @@
...
+          .clickProjectBeSharedYes()
+          .clicksignedaNDANo()
+          .enterLandMonth()
+          .enterLandYear()
+      },
+      submitTheForm () {
+        return this
+        .submitForm('form')
+      },
+      createNewInvestmentProjectWithDifferentReferralSourceAdviser (projectName) {
```
Comment (tyom) [inline test/acceptance/page_objects/InvestmentProject.js]: Adding new line before this will make it easier to distinguish and would be consistent with the code above.

File: test/acceptance/features/step_definitions/create-a-investment-project-step-definitions.js [hunk 18]
```diff
@@ -0,0 +1,224 @@
...
+  When(/^I create a new Investment project with different client relation manager and referral source adviser$/, async () => {
+    projectName = faker.commerce.productName()
+    await Company
+      .navigate()
+      .findCompany(randomCompanyName)
+    await Contact
+      .clickOnFirstCompanyFromList()
+    await Investment
+      .clickInvestmentsTab()
+      .createNewInvestmentProjectWithDifferentClientRelationManagerAndReferralSourceAdviser(projectName)
```
Comment (tyom) [inline test/acceptance/features/step_definitions/create-a-investment-project-step-definitions.js]: Quite a mouthful.

File: test/acceptance/features/step_definitions/create-a-investment-project-step-definitions.js (hunk 2 above)
Comment (venureddy95) [inline test/acceptance/features/step_definitions/create-a-investment-project-step-definitions.js]: @feedmypixel : Changed it to single quotes

File: test/acceptance/features/step_definitions/create-a-investment-project-step-definitions.js (hunk 3 above)
Comment (venureddy95) [inline test/acceptance/features/step_definitions/create-a-investment-project-step-definitions.js]: @feedmypixel : Changed it to single quotes

File: test/acceptance/features/step_definitions/create-a-investment-project-step-definitions.js (hunk 4 above)
Comment (venureddy95) [inline test/acceptance/features/step_definitions/create-a-investment-project-step-definitions.js]: @feedmypixel : `projectName` is set dynamically in each step definition `When` , to keep it unique for each test. I am removing the blank declaration for now.

File: test/acceptance/features/step_definitions/create-a-investment-project-step-definitions.js (hunk 5 above)
Comment (venureddy95) [inline test/acceptance/features/step_definitions/create-a-investment-project-step-definitions.js]: @feedmypixel : Implemented the arrow function in all places.

File: test/acceptance/features/step_definitions/create-a-investment-project-step-definitions.js (hunk 6 above)
Comment (venureddy95) [inline test/acceptance/features/step_definitions/create-a-investment-project-step-definitions.js]: @feedmypixel : Changed it to `typeOfInvestment`

File: test/acceptance/page_objects/InvestmentProject.js (hunk 7 above)
Comment (venureddy95) [inline test/acceptance/page_objects/InvestmentProject.js]: @feedmypixel : Removed it.

File: test/acceptance/page_objects/InvestmentProject.js (hunk 8 above)
Comment (venureddy95) [inline test/acceptance/page_objects/InvestmentProject.js]: @feedmypixel This would be a great idea to add ID hooks or class names to the form elements that I can use for Automation. Please feel free to update the non-definitive elements that I used in Automation to a more definitive ones, or let me know them, i am happy to update across the Page Objects.

File: test/acceptance/page_objects/InvestmentProject.js (hunk 9 above)
Comment (venureddy95) [inline test/acceptance/page_objects/InvestmentProject.js]: @feedmypixel: removed it now.

File: test/acceptance/page_objects/InvestmentProject.js (hunk 11 above)
Comment (venureddy95) [inline test/acceptance/page_objects/InvestmentProject.js]: @feedmypixel @tyom : I agree with one line abstractions are not that useful. All I wanted to achieve here is just a naming convention for each method, which i will be using in the actual file that will read around 10 methods. In future if any elements have changed, then I can refer to the method name and understand what it is referring to and make necessary changes. I can discuss this further in office and make a conclusion.

File: test/acceptance/features/step_definitions/create-a-investment-project-step-definitions.js (hunk 12 above)
Comment (venureddy95) [inline test/acceptance/features/step_definitions/create-a-investment-project-step-definitions.js]: @tyom : Changed.

File: test/acceptance/features/step_definitions/create-a-investment-project-step-definitions.js (hunk 13 above)
Comment (venureddy95) [inline test/acceptance/features/step_definitions/create-a-investment-project-step-definitions.js]: @tyom : Changed.

File: test/acceptance/features/step_definitions/create-a-investment-project-step-definitions.js (hunk 14 above)
Comment (venureddy95) [inline test/acceptance/features/step_definitions/create-a-investment-project-step-definitions.js]: @tyom : Changed.

File: test/acceptance/page_objects/InvestmentProject.js (hunk 16 above)
Comment (venureddy95) [inline test/acceptance/page_objects/InvestmentProject.js]: @tyom : Changed.

File: test/acceptance/page_objects/InvestmentProject.js (hunk 17 above)
Comment (venureddy95) [inline test/acceptance/page_objects/InvestmentProject.js]: @tyom : Added new line.

File: test/acceptance/page_objects/InvestmentProject.js (hunk 15 above)
Comment (venureddy95) [inline test/acceptance/page_objects/InvestmentProject.js]: @tyom : Changed.

File: test/acceptance/features/step_definitions/create-a-investment-project-step-definitions.js (hunk 18 above)
Comment (venureddy95) [inline test/acceptance/features/step_definitions/create-a-investment-project-step-definitions.js]: @tyom : Changed it to `createNewInvestmentProjectWithDifferentClientAndReferralDetails`

File: test/acceptance/page_objects/InvestmentProject.js (hunk 11 above)
Comment (feedmypixel) [inline test/acceptance/page_objects/InvestmentProject.js]: Could you not just do that with find/replace?

Comment (feedmypixel) [inline test/acceptance/page_objects/InvestmentProject.js]: Agreed - Lets catch up face to face on this
